### DISCUSSION TITLE
Language Specific Leaderboards

### DISCUSSION THREAD

**Discussion by Kacarott** [codewars/codewars.com#2371] [Ideas]
Another simple idea, create a (perhaps only top 100) leaderboard for each language. There could be one single "Language leaderboards" page with a dropdown menu selecting which.

Benefits:
- Nice for users specialising in 'niche' languages, who otherwise don't complete enough kata to make it on the normal leaderboard
- Interesting to see who the top users in the community are for each language
- Might encourage more people to try the beta/barely touched languages, since getting on the leaderboard would be 'easier'.

**Comment by Kacarott** [comment]
As an aside (going along with the "Badge" idea from the other post), a badge showing leaderboard ranking for a specific language  would *really* encourage people to try out the barely touched katas, just to be among the top 10 or so users in *any* language.

**Reply by ejini6969** [reply]
Another benefit of "language leaderboards" is that it will encourage some users to make new translations since doing so will grant them language points. ~~

**Comment by hobovsky** [comment]
I was thinking about having more rank-related rankings and leaderboards in general. Honor leaderboards reflect mostly users who grind or prefer quantity over quality. Solving many easy kata can bring you quite far on honor leaderboards.

I would love to have something what would appreciate users preferring difficult tasks. It could be, for example:
- Overall rank score leaderboard
- Language leaderboards
- Something similar to [Olympic medal table](https://en.wikipedia.org/wiki/Olympic_medal_table), for example gold first, or some weighted ranking

Language leaderboards could additionally help finding users who'd be potentially helpful with reviewing, approving, or fixing translations or kata in particular languages, or could be used to drive some potential maintenance/approval privileges in the future.

**Reply by paulo-techie** [reply]
I don't think so. I have only done 37 katas and I have risen quickly by choosing questions with high-ranked kyus. So if you want to rise doing many easy katas then I see no problem with that too. There is still a good reward for doing tougher challenges than for doing many easy katas. I think it's fair so far from what I have experienced.

**Comment by hobovsky** [comment]
Potential problem I see with language leaderboards: criteria for awarding language progress are somewhat messy and potentially bugged. You gain language progress when you first solve a kata, and when you solve it afterwards in another language. But as soon as you get marked as a contributor, your ability for gaining language progress gets hindered in a somewhat unpredictable (or at least undocumented) way. This makes it difficult to max out your language rank in any way other than only solving kata available in language of your choice, and generally stops you from any authoring and maintenance activities.

**Reply by kazk** [reply]
> But as soon as you get marked as a contributor, your ability for gaining language progress gets hindered in a somewhat unpredictable (or at least undocumented) way. 

I want to change that, and undoing the effects should be possible too. The number of affected users is relatively low.

**Comment by kazk** [comment]
More than 4 years ago, I wrote the following as a Codewars user (creating an external leaderboard):

> I was thinking to add some leaderboards not based on honor. Here're some ideas I got after looking at data from `/users/:user` of top 50 users on "Completed Kata" leaderboard:
> 
> - `ranks.overall.score`
> - `ranks.languages.{language}.score` for popular languages
>     - shows experts in that language
>     - most in top 50 were polyglots, found 1 user who did only Python
> - sum of `ranks.languages.{language}.score`
>     - polyglots who solves the same kata in multiple languages
>     - makes me the 1st :smirk:
> - `codeChallenges.totalCompleted`
>     - `4096`!
> - `codeChallenges.totalAuthored`
>     - `682`!
> - number of approved translations
>     - I couldn't find a way to get this
>
> https://github.com/codewars/codewars-runner-cli/issues/344#issuecomment-291379388

**Reply by hobovsky** [reply]
+1 for polyglot board

**Comment by DanielRomanMartinez** [comment]
Can we see it in Codewars website? Would be nice to see it :)

**Comment by kazk** [comment]
For now, you can use the API (`/api/v1/leaders/ranks/:language`) to find leaders:

- https://www.codewars.com/api/v1/leaders/ranks/overall
- https://www.codewars.com/api/v1/leaders/ranks/javascript

No UI yet because it's _inaccurate_ due to a known issue where contributors are not gaining rank scores after contributing. A fix for that is in progress.

**Reply by thekernal** [reply]
Awesome, I was looking for this because I primarily program in Perl, and only a subset of katas are converted for it.  I could easily see where I stood can't wait to have a UI for it. Good luck fixing contributors.

**Reply by hobovsky** [reply]
If you want some UI for the feature despite it not being completely accurate, there are some options:

A [userscript](https://github.com/hobovsky/polyglot) (created by me) which shows leaderboards in CW UI:

![image](https://user-images.githubusercontent.com/23709795/187922141-ed2f8b82-3450-4578-ac14-4103f3e56a5b.png)

A `/leaderboard` command on Codewars Discord:

![image](https://user-images.githubusercontent.com/23709795/187922521-0ee07f34-33f2-4ecd-a94d-762af5dafa76.png)

I guess the in-site UI will be available after the contributor flag gets fixed.

**Comment by kazk** [comment]
Deployed https://www.codewars.com/users/leaderboard/ranks

**Comment by jon-torlen** [comment]
exelente
